This will shorten the list of allowed currencies

melaxon opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->
This will shorten the list of allowed currencies showing only installed currencies
### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
in Store->Configuration->General->Currency setup will be shown only those currencies selected in 
Installed currencies list (Configuration->Advanced->System)
### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. magento/magento2#<issue_number>: Issue title
2. ...

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. Currently if you select any currency in the list in Store->Configuration->General->Currency setup and this currency is not yet selected in Installed currency list (Configuration->Advanced->System) you will receive error message. 
2. After the modification is applied there will be only those currencies in the list (Store->Configuration->General->Currency setup) that can be selected without any error message

### Contribution checklist (*)
 - [ x] Pull request has a meaningful description of its purpose
 - [ x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

melaxon committed:
This will shorten the list of allowed currencies

in Store->Configuration->General->Currency setup will be shown only those currencies selected in 
Installed currencies list (Configuration->Advanced->System)

magento-engcom-team commented:
Hi @melaxon. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

bartoszkubicki commented on lib/internal/Magento/Framework/Locale/TranslatedLists.php:
```diff
@@ -150,10 +157,20 @@ public function getOptionCurrencies()
         $allowed = $this->_config->getAllowedCurrencies();
+        $selectedCurrencies = explode(
+            ',', $this->scopeConfig->getValue(
+                'system/currency/installed',
```
I would put inside private const inside class

orlangur replied:
Not needed until it is used only once.

bartoszkubicki commented on lib/internal/Magento/Framework/Locale/TranslatedLists.php:
```diff
@@ -150,10 +157,20 @@ public function getOptionCurrencies()
+        $selectedCurrencies = explode(
+            ',', $this->scopeConfig->getValue(
+                'system/currency/installed',
+                \Magento\Store\Model\ScopeInterface::SCOPE_STORE
```
Maybe it could be move to imports and here left in short version

bartoszkubicki commented on lib/internal/Magento/Framework/Locale/TranslatedLists.php:
```diff
@@ -150,10 +157,20 @@ public function getOptionCurrencies()
             if (!in_array($code, $allowed)) {
                 continue;
             }
+            if (!in_array($code, $selectedCurrencies)) {
```
I think condition could be moved to separate protected function, in order to encapsulate it

bartoszkubicki commented on lib/internal/Magento/Framework/Locale/TranslatedLists.php:
```diff
@@ -150,10 +157,20 @@ public function getOptionCurrencies()
         $currencies = (new CurrencyBundle())->get($this->localeResolver->getLocale())['Currencies'] ?: [];
         $options = [];
         $allowed = $this->_config->getAllowedCurrencies();
+        $selectedCurrencies = explode(
```
Maybe getting selecting countries could be done in separate protected method?

melaxon replied:
This function will be called only once. Does it deserve it?

sivaschenko commented:
Hi @melaxon Thanks for the contribution! Please see the code review notes and consider covering the introduced improvement with a test. Also failed unit tests need to be fixed.

melaxon commented:
@magento-engcom-team give me test instance

magento-engcom-team commented:
Hi @melaxon. Thank you for your request. I'm working on Magento instance for you

magento-engcom-team commented:
Hi @melaxon, here is your [new Magento instance](https://pr-19946.instances.magento-community.engineering).
Admin access: [https://pr-19946.instances.magento-community.engineering/admin](https://pr-19946.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

melaxon commented:
It works in test instance.
I allowed UDS and EUR in Configuration->Advanced->System
Then I see only allowed currencies in General->Currency Setup:
https://gyazo.com/c6a970fd6d34d1459ee7e9ed47828822

melaxon commented:
Trying to solve this:
TypeError: Argument 1 passed to Magento\Framework\Locale\TranslatedLists::__construct() must implement interface Magento\Framework\App\Config\ScopeConfigInterface, instance of Mock_ConfigInterface_33ac175b given, called in /home/travis/build/magento/magento2/lib/internal/Magento/Framework/Locale/Test/Unit/TranslatedListsTest.php on line 40

melaxon committed:
Trying to keep The Travis happy

Trying to solve this:
TypeError: Argument 1 passed to Magento\Framework\Locale\TranslatedLists::__construct() must implement interface Magento\Framework\App\Config\ScopeConfigInterface, instance of Mock_ConfigInterface_33ac175b given, called in /home/travis/build/magento/magento2/lib/internal/Magento/Framework/Locale/Test/Unit/TranslatedListsTest.php on line 40

melaxon committed:
fix test unit

melaxon committed:
fix test unit

fix test unit

melaxon commented:
done

sivaschenko commented on lib/internal/Magento/Framework/Locale/TranslatedLists.php:
```diff
@@ -13,6 +13,11 @@
+    /**
+     * @var TranslatedLists
+     */
+    public $selectedCurrencies;
```
Please change the access level of new properties to private

sivaschenko commented on lib/internal/Magento/Framework/Locale/Test/Unit/TranslatedListsTest.php:
```diff
@@ -64,12 +73,15 @@ public function testGetOptionAllCurrencies()
             ->will($this->returnValue($allowedCurrencies));
 
-        $expectedResults = ['USD', 'EUR', 'GBP', 'UAH'];
+        $this->listsModel->selectedCurrencies = $selectedCurrencies;
```
Please mock the return of selected currencies from the `scopeConfig`.

melaxon replied:
ok. Along with the above comment

sivaschenko commented on lib/internal/Magento/Framework/Locale/Test/Unit/TranslatedListsTest.php:
```diff
@@ -64,12 +73,15 @@ public function testGetOptionAllCurrencies()
     {
         $allowedCurrencies = ['USD', 'EUR', 'GBP', 'UAH'];
-
+		$selectedCurrencies = ['USD', 'EUR'];
```
```suggestion
        $selectedCurrencies = ['USD', 'EUR'];
```

sivaschenko reviewed (changes requested):
Hi @melaxon please see the code review notes

melaxon commented:
I understood that the Framework is not the place where such modification can be applied because it is based on user input. New PR will be opened now
